File: mailblastr-php/src/Resources/Campaigns.php
<?php

declare(strict_types=1);

namespace Mailblastr\Resources;

use Mailblastr\Client;

/**
 * Campaigns — bulk sends to a domain's contact pool (or a segment of it).
 * DOMAIN-FIRST: 'domain' is REQUIRED on create and picks the contact pool the
 * campaign targets (orthogonal to 'from', which may be a different verified
 * domain).
 */
class Campaigns extends Resource
{
    /**
     * Create a campaign. POST /campaigns
     *
     * @param array $payload domain (REQUIRED), from, subject, html/text,
     *                       reply_to, preview_text, name, segment_id, topic_id,
     *                       recurrence, recurrence_every, ab_test, followups,
     *                       list_to, unsubscribe_policy, send, scheduled_at,
     *                       schedule_timezone (IANA zone the schedule + daily
     *                       batching are evaluated in), daily_batch_size (max
     *                       recipients per batch-day, 1-100000).
     */
    public function create(array $payload): array
    {
        return $this->client->request('POST', '/campaigns', $payload);
    }

    /** Retrieve a campaign (includes statistics). GET /campaigns/:id */
    public function get(string $id): array
    {
        return $this->client->request('GET', '/campaigns/' . Client::e($id));
    }

    /**
     * List campaigns. GET /campaigns
     *
     * @param array $params Cursor pagination: limit, after, before.
     */
    public function list(array $params = []): array
    {
        return $this->client->request('GET', '/campaigns' . $this->paginationQuery($params));
    }

    /**
     * Update a draft campaign. PATCH /campaigns/:id
     *
     * @param array $payload The create-side fields plus followups (replace
     *                       pending follow-ups; [] clears), list_to (bool),
     *                       unsubscribe_policy ('account'|'domain'|'ignore'),
     *                       schedule_timezone (IANA zone; null clears), and
     *                       daily_batch_size (1-100000; null clears).
     */
    public function update(string $id, array $payload): array
    {
        return $this->client->request('PATCH', '/campaigns/' . Client::e($id), $payload);
    }

    /**
     * Send now, or schedule with ['scheduled_at' => …]. An optional
     * 'schedule_timezone' (IANA name) is persisted onto the campaign so daily
     * batching evaluates batch-days in that zone. POST /campaigns/:id/send
     */
    public function send(string $id, array $payload = []): array
    {
        return $this->client->request('POST', '/campaigns/' . Client::e($id) . '/send', $payload);
    }

    /** Cancel a scheduled campaign (returns it to draft). POST /campaigns/:id/cancel */
    public function cancel(string $id): array
    {
        return $this->client->request('POST', '/campaigns/' . Client::e($id) . '/cancel');
    }

    /** Per-campaign analytics (counts, engagement rates, top 50 links). GET /campaigns/:id/stats */
    public function stats(string $id): array
    {
        return $this->client->request('GET', '/campaigns/' . Client::e($id) . '/stats');
    }

    /**
     * Who opened, clicked and replied — the per-contact rows behind
     * {@see stats()}. Not paginated; each list is capped at 500 rows.
     * GET /campaigns/:id/engagement
     */
    public function engagement(string $id): array
    {
        return $this->client->request('GET', '/campaigns/' . Client::e($id) . '/engagement');
    }

    /** A/B winner evaluation for an A/B campaign. GET /campaigns/:id/ab */
    public function ab(string $id): array
    {
        return $this->client->request('GET', '/campaigns/' . Client::e($id) . '/ab');
    }

    /** Delete a campaign. DELETE /campaigns/:id */
    public function remove(string $id): array
    {
        return $this->client->request('DELETE', '/campaigns/' . Client::e($id));
    }
}

File: mailblastr-ruby/lib/mailblastr/campaigns.rb
# frozen_string_literal: true

module Mailblastr
  # Campaigns are DOMAIN-FIRST: `domain` (required on create) picks the
  # contact pool the campaign targets; `from` may be any verified domain.
  module Campaigns
    class << self
      # POST /campaigns
      #   Mailblastr::Campaigns.create({ domain: "yourdomain.com", from: "you@yourdomain.com",
      #                                  subject: "Hello", html: "<p>Hi</p>", segment_id: "seg_1" })
      # Optional scheduling params: `schedule_timezone` (IANA zone the schedule +
      # daily batching are evaluated in, e.g. "America/New_York") and
      # `daily_batch_size` (max recipients per batch-day, 1-100000).
      def create(params)
        Client.require_domain!(params, "Campaigns.create")
        Client.request(:post, "/campaigns", body: params)
      end

      # GET /campaigns/:id
      def get(campaign_id)
        Client.request(:get, "/campaigns/#{Client.path_escape(campaign_id)}")
      end

      # GET /campaigns
      def list(params = {})
        Client.request(:get, "/campaigns", query: Client.pagination(params))
      end

      # PATCH /campaigns/:id (draft campaigns only). Accepts the create-side
      # fields plus `followups` (replace pending follow-ups, [] clears),
      # `list_to` (true/false), `unsubscribe_policy` ("account" | "domain" |
      # "ignore"), `schedule_timezone` (IANA zone; nil clears), and
      # `daily_batch_size` (1-100000; nil clears).
      def update(campaign_id, params)
        Client.request(:patch, "/campaigns/#{Client.path_escape(campaign_id)}", body: params)
      end

      # Send now, or schedule with { scheduled_at: "..." }. An optional
      # `schedule_timezone` (IANA name) is persisted onto the campaign so daily
      # batching evaluates batch-days in that zone. POST /campaigns/:id/send
      def send(campaign_id, params = {})
        Client.request(:post, "/campaigns/#{Client.path_escape(campaign_id)}/send", body: params)
      end

      # Stop a campaign's remaining work. Accepted on `scheduled`, `recurring`,
      # `paused` and `queued`; anything else is a validation_error.
      # `scheduled`/`recurring`/`paused` return to `draft` (editable, re-sendable);
      # a `queued` campaign already fanning out becomes the TERMINAL `canceled`,
      # which can never be edited, re-sent or deleted. Copies already handed to
      # the mail service cannot be recalled; what stops is every remaining
      # recipient (for a staggered campaign, every future batch-day).
      # Read the returned `status` to see which happened.
      # POST /campaigns/:id/cancel
      def cancel(campaign_id)
        Client.request(:post, "/campaigns/#{Client.path_escape(campaign_id)}/cancel")
      end

      # Per-campaign analytics (counts, engagement rates, top links). GET /campaigns/:id/stats
      def stats(campaign_id)
        Client.request(:get, "/campaigns/#{Client.path_escape(campaign_id)}/stats")
      end

      # Who opened, clicked and replied, contact by contact. Each list is
      # capped at 500 rows and there is no pagination.
      # GET /campaigns/:id/engagement
      def engagement(campaign_id)
        Client.request(:get, "/campaigns/#{Client.path_escape(campaign_id)}/engagement")
      end

      # A/B winner evaluation for an A/B campaign. GET /campaigns/:id/ab
      def ab(campaign_id)
        Client.request(:get, "/campaigns/#{Client.path_escape(campaign_id)}/ab")
      end

      # DELETE /campaigns/:id
      def delete(campaign_id)
        Client.request(:delete, "/campaigns/#{Client.path_escape(campaign_id)}")
      end
    end
  end
end
